chains/ethereum/chain.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package ethereum
+
+import (
+	bridge "github.com/ChainSafe/ChainBridge/bindings/Bridge"
+	erc20Handler "github.com/ChainSafe/ChainBridge/bindings/ERC20Handler"
+	"github.com/ChainSafe/ChainBridge/chains"
+	"github.com/ChainSafe/ChainBridge/core"
+	"github.com/ChainSafe/ChainBridge/crypto/secp256k1"
+	"github.com/ChainSafe/ChainBridge/keystore"
+	msg "github.com/ChainSafe/ChainBridge/message"
+	"github.com/ChainSafe/ChainBridge/router"
+	"github.com/ChainSafe/log15"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type Chain struct {
+	cfg      *core.ChainConfig // The config of the chain
+	conn     *Connection       // THe chains connection
+	listener *Listener         // The listener of this chain
+	writer   *Writer           // The writer of the chain
+}
+
+func createBridgeContract(addr common.Address, conn *Connection) (*BridgeContract, error) {
+	bridgeInstance, err := bridge.NewBridge(addr, conn.conn)
+	if err != nil {
+		return nil, err
+	}
+
+	raw := &bridge.BridgeRaw{
+		Contract: bridgeInstance,
+	}
+
+	return &BridgeContract{
+		BridgeRaw:        raw,
+		BridgeCaller:     &bridgeInstance.BridgeCaller,
+		BridgeTransactor: &bridgeInstance.BridgeTransactor,
+	}, nil
+}
+
+func createErc20HandlerContract(addr common.Address, conn *Connection) (*ERC20HandlerContract, error) {
+	instance, err := erc20Handler.NewERC20Handler(addr, conn.conn)
+	if err != nil {
+		return nil, err
+	}
+
+	raw := &erc20Handler.ERC20HandlerRaw{
+		Contract: instance,
+	}
+
+	return &ERC20HandlerContract{
+		ERC20HandlerRaw:    raw,
+		ERC20HandlerCaller: &instance.ERC20HandlerCaller,
+	}, nil
+}
+
+func InitializeChain(chainCfg *core.ChainConfig) (*Chain, error) {
+	cfg, err := parseChainConfig(chainCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	kpI, err := keystore.KeypairFromAddress(cfg.from, keystore.EthChain, cfg.keystorePath, chainCfg.Insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	kp, _ := kpI.(*secp256k1.Keypair)
+
+	logger := log15.Root().New("chain", cfg.name)
+
+	conn := NewConnection(cfg, kp, logger)
+	err = conn.Connect()
+	if err != nil {
+		return nil, err
+	}
+
+	err = conn.ensureHasBytecode(cfg.bridgeContract)
+	if err != nil {
+		return nil, err
+	}
+	err = conn.ensureHasBytecode(cfg.erc20HandlerContract)
+	if err != nil {
+		return nil, err
+	}
+	err = conn.ensureHasBytecode(cfg.genericHandlerContract)
+	if err != nil {
+		return nil, err
+	}
+
+	bridgeContract, err := createBridgeContract(cfg.bridgeContract, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	erc20HandlerContract, err := createErc20HandlerContract(cfg.erc20HandlerContract, conn)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Why don't we initialize the generic contract?

Reply:
Updated in #320 (#323)